pkg/clientutil/crds.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2020 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clientutil
+
+import (
+	"context"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	dynclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CRDsReadyCondition generate a k8s.io/apimachinery/pkg/util/wait.ConditionFunc function to be used in
+// k8s.io/apimachinery/pkg/util/wait.Poll* family of functions. It will check all provided GKs (GroupKinds) to exists
+// and have Established status
+func CRDsReadyCondition(ctx context.Context, client dynclient.Client, groupKinds []string) func() (bool, error) {
+	return func() (bool, error) {
+		for _, gk := range groupKinds {
+			crd := apiextensions.CustomResourceDefinition{}
+			key := dynclient.ObjectKey{Name: gk}
+
+			if err := client.Get(ctx, key, &crd); err != nil {
+				return false, err
+			}
+
+			for _, cond := range crd.Status.Conditions {
+				if cond.Type == apiextensions.Established && cond.Status != apiextensions.ConditionTrue {
+					return false, nil
+				}
```

Comment:
I was thinking about this approach and I'm not sure that it's safe at all.

This check assumes that the Established condition always exists on the CRD object, but I'm not sure is that true. If `client.Get` returns the result before CRD controller sets the initial conditions, it maybe could happen that `crd.Status.Conditions` don't contain any Established condition. In that case, this check would pass and the function would return true, while the CRD is not Established.

I believe that the following check would be safer

```suggestion
				if cond.Type == apiextensions.Established && cond.Status == apiextensions.ConditionTrue {
					return true, nil
				}
```

It needs some other changes, as well as, it should be tested.

Reply:
Changing it to such form means introducing a bug. We want to check all of CRDs, but in this form the first found CRD will report positive on behalf of all CRDs under test.

Reply:
In our case, I believe something like this would work and not introduce a bug:
```go
var found bool
for _, cond := range crd.Status.Conditions {
    if cond.Type == apiextensions.Established && cond.Status == apiextensions.ConditionTrue {
        found = true
        break
    }
}
if !found {
    return false, nil
}
```

Reply:
yeah, changed to something similar